### PR TITLE
Feature/adl 32

### PR DESCRIPTION
**Pull request by Sam6wani**
Set up Connection page related changes.

### CODE REVIEW

File: Dashboard/App.axaml.cs [hunk 1]
```diff
@@ -27,12 +27,19 @@ public override void OnFrameworkInitializationCompleted()
             // Setup dependency injection
             var services = DependencyInjection.ConfigureServices();
 
-            desktop.MainWindow = new MainWindow
+            /*desktop.MainWindow = new MainWindow
```
Comment (nrdag) [inline Dashboard/App.axaml.cs]: I know that these windows won't be displayed like this in the future, but for the purpose of this ticket you can just replace this line with
```suggestion
            desktop.MainWindow = new ConnectionWindow();
```

File: Dashboard/Views/ConnectionWindow.axaml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+        x:Class="Dashboard.ConnectionWindow"
+        Title="ConnectionWindow"
+        Width="800"
+        Height="500">
+    <DockPanel>
+
+        <!-- Top Menu -->
+        <Menu DockPanel.Dock="Top" Background="LightGray">
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml]: The menu isn't required here, richard's ticket will add this at a later date.

Comment (nrdag) [reply Dashboard/Views/ConnectionWindow.axaml]: also the bottom status panel not required in this ticket

File: Dashboard/Views/ConnectionWindow.axaml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            <Grid.ColumnDefinitions>
+                <ColumnDefinition Width="2*"/>
+                <ColumnDefinition Width="3*"/>
+            </Grid.ColumnDefinitions>
+
+            <StackPanel Grid.Column="0" Margin="20">
+                <!-- Connection Type -->
+                <TextBlock Text="Connection Type" Margin="0, 10, 0, 5"/>
+                <ComboBox Width="200">
+                    <ComboBoxItem>Serial Port</ComboBoxItem>
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml]: This should be only:
- Serial port
- TCP
- File (CSV maybe?)

File: Dashboard/Views/ConnectionWindow.axaml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                    <ComboBoxItem>Ethernet</ComboBoxItem>
+                    <ComboBoxItem>USB</ComboBoxItem>
+                    <ComboBoxItem>Wi-Fi</ComboBoxItem>
+                </ComboBox>
+
+                <!-- Serial Port -->
+                <TextBlock Text="Serial Port" Margin="0, 10, 0, 5"/>
+                <ComboBox Width="200">
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml]: These serial ports should be discovered dynamically, by listing the available ports from the computer

File: Dashboard/Views/ConnectionWindow.axaml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                    <ComboBoxItem>COM1</ComboBoxItem>
+                    <ComboBoxItem>COM2</ComboBoxItem>
+                    <ComboBoxItem>COM3</ComboBoxItem>
+                    <ComboBoxItem>COM4</ComboBoxItem>
+                    <ComboBoxItem>COM5</ComboBoxItem>
+                </ComboBox>
+
+                <!-- IP Address -->
+                <TextBlock Text="IP Address" Margin="0, 15, 0, 5"/>
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml]: Each of these sections should be hidden until the connection type is selected above, e.g. if serial port is selected hide everything besides the serial port drop down.

File: Dashboard/Views/ConnectionWindow.axaml.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    private void InitializeComponent()
+    {
+        AvaloniaXamlLoader.Load(this);
+    }
+
+    protected override void OnClosed(EventArgs e)
+    {
+        base.OnClosed(e);
+
+        // Dispose the view model when the window is closed, to clean up the data store and connector
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml.cs]: Remove commented out code

**Review by nrdag — CHANGES_REQUESTED** [review]
Please have a look through at my comments for my suggested changes.

File: Dashboard/Views/ConnectionWindow.axaml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+                <!-- Simulation -->
+                <TextBlock Text="Simulation" Margin="0, 10, 0, 5"/>
+                <StackPanel Orientation="Horizontal">
+                    <TextBox Width="200"/>
+                    <Button Content="..." Width="30" Margin="5,0,0,0"/>
+                </StackPanel>
+
+                <!-- Connect Button -->
+                <Button Content="Connect" Width="100" Margin="0,30,0,0" Background="#00FFFF" FontWeight="Bold"/>
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml]: Please remove the width attribute here, not required (it makes the button longer than needed).

According to AC, this button should be disabled / greyed out for now.

**Comment by Sam6wani** [comment]
I fixed all. Please review now.

File: Dashboard/ViewModels/ConnectionWindowViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+using ReactiveUI;
+using System;
+using System.ComponentModel;
+
+namespace Dashboard.ViewModels 
+{
+    internal class ConnectionWindowViewModel : ReactiveObject
+    {
+
+        /*private string _selectedConnectionType;
```
Comment (nrdag) [inline Dashboard/ViewModels/ConnectionWindowViewModel.cs]: Why is all of this code commented out?

File: Dashboard/Dashboard.csproj [hunk 9]
```diff
@@ -9,17 +9,24 @@
...
-        <AvaloniaResource Include="Assets\**"/>
+        <AvaloniaResource Include="Assets\**" />
     </ItemGroup>
 
     <ItemGroup>
-        <PackageReference Include="Avalonia" Version="11.1.0"/>
-        <PackageReference Include="Avalonia.Desktop" Version="11.1.0"/>
-        <PackageReference Include="Avalonia.Themes.Fluent" Version="11.1.0"/>
-        <PackageReference Include="Avalonia.Fonts.Inter" Version="11.1.0"/>
+      <None Remove="Assets\refresh.png" />
```
Comment (nrdag) [inline Dashboard/Dashboard.csproj]: This file isn't used, why was it added?

File: Dashboard/App.axaml.cs [hunk 10]
```diff
@@ -27,12 +27,20 @@ public override void OnFrameworkInitializationCompleted()
             // Setup dependency injection
             var services = DependencyInjection.ConfigureServices();
 
-            desktop.MainWindow = new MainWindow
+          /*  desktop.MainWindow = new MainWindow
             {
                 DataContext = services.GetRequiredService<MainViewModel>()
+            };*/
+
+            desktop.MainWindow = new ConnectionWindow()
```
Comment (nrdag) [inline Dashboard/App.axaml.cs]: Can you please merge in the main branch, and adjust your code to work with the connection tab?

File: Dashboard/Dashboard.csproj [hunk 11]
```diff
@@ -9,17 +9,24 @@
...
+    </ItemGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Avalonia" Version="11.1.0" />
+        <PackageReference Include="Avalonia.Desktop" Version="11.1.0" />
+        <PackageReference Include="Avalonia.ReactiveUI" Version="11.1.0" />
```
Comment (nrdag) [inline Dashboard/Dashboard.csproj]: This package isnt used

File: Dashboard/Dashboard.csproj [hunk 12]
```diff
@@ -9,17 +9,24 @@
...
+        <PackageReference Include="Avalonia.Themes.Fluent" Version="11.1.0" />
+        <PackageReference Include="Avalonia.Fonts.Inter" Version="11.1.0" />
         <!--Condition below is needed to remove Avalonia.Diagnostics package from build output in Release configuration.-->
-        <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.1.0"/>
-        <PackageReference Include="CommunityToolkit.Mvvm" Version="8.2.1"/>
-        <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0"/>
+        <PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.1.0" />
+        <PackageReference Include="CommunityToolkit.Mvvm" Version="8.2.1" />
+        <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0" />
+        <PackageReference Include="ReactiveUI" Version="20.1.52" />
```
Comment (nrdag) [inline Dashboard/Dashboard.csproj]: Also not used

File: Dashboard/Views/ConnectionWindow.axaml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+            <Grid.ColumnDefinitions>
+                <ColumnDefinition Width="2*"/>
+                <ColumnDefinition Width="3*"/>
+            </Grid.ColumnDefinitions>
+
+            <StackPanel Grid.Column="0" Margin="20">
+                <!-- Connection Type -->
+                <TextBlock Text="Connection Type" Margin="0, 10, 0, 5"/>
+                <ComboBox Width="200">
+                    <ComboBoxItem>Serial Port</ComboBoxItem>
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml]: Can you please make it so that the other fields are hidden unless selected, can you also make it so this combo box defaults to the first value (Serial Port)? -- Current it is initially blank

**Review by nrdag — CHANGES_REQUESTED** [review]
Added some comments

File: Dashboard/Views/ConnectionWindow.axaml.cs [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+    }
+
+    protected override void OnClosed(EventArgs e)
+    {
+        base.OnClosed(e);
+
+    }
+
+    private void RefreshPorts(object? sender, RoutedEventArgs? e)
```
Comment (S3784866) [inline Dashboard/Views/ConnectionWindow.axaml.cs]: Adding onto Nicks comments, this function currently adds on top of what is already present in the ComboBox resulting in a expanding list of duplicate values when refreshed multiple times. Please adjust.
connectionWindowComboBox

File: Dashboard/ViewModels/ConnectionWindowViewModel.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+using System.ComponentModel;
+
+namespace Dashboard.ViewModels 
+{
+    internal class ConnectionWindowViewModel : ReactiveObject
+    {
+
+        private string _selectedConnectionType;
+
+        public string SelectedConnectionType
```
Comment (nrdag) [inline Dashboard/ViewModels/ConnectionWindowViewModel.cs]: This probably shouldn't be a string, we can only select 3 options. why not use an enum?

Also, you can replace almost this entire class with just
`[ObservableProperty] public string _selectedConnectionType = string.Empty;`

File: Dashboard/ViewModels/ConnectionWindowViewModel.cs [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+        {
+            get => _selectedConnectionType;
+            set => this.RaiseAndSetIfChanged(ref _selectedConnectionType, value);
+        }
+
+        public ConnectionWindowViewModel()
+        {
+            // Default to no selection
```
Comment (nrdag) [inline Dashboard/ViewModels/ConnectionWindowViewModel.cs]: Default to serial port please

File: Dashboard/Dashboard.csproj (hunk 9 above)
Comment (nrdag) [reply Dashboard/Dashboard.csproj]: this file still isnt used

File: Dashboard/Dashboard.csproj (hunk 11 above)
Comment (nrdag) [reply Dashboard/Dashboard.csproj]: after changing the view model code, this package isnt needed anymore

File: Dashboard/Dashboard.csproj (hunk 12 above)
Comment (nrdag) [reply Dashboard/Dashboard.csproj]: after changing the view model code, this package isnt needed anymore

File: Dashboard/Program.cs [hunk 17]
```diff
@@ -1,4 +1,5 @@
 ﻿using Avalonia;
+using Avalonia.ReactiveUI;
```
Comment (nrdag) [inline Dashboard/Program.cs]: Why is this here, not used please remove.

File: DashboardTests/DashboardTests.csproj [hunk 18]
```diff
@@ -10,18 +10,19 @@
...
-        <PackageReference Include="Avalonia.Headless.NUnit" Version="11.1.3"/>
-        <PackageReference Include="Avalonia.Themes.Fluent" Version="11.1.3"/>
-        <PackageReference Include="coverlet.collector" Version="6.0.0"/>
-        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0"/>
-        <PackageReference Include="Moq" Version="4.20.70"/>
-        <PackageReference Include="NUnit" Version="3.14.0"/>
-        <PackageReference Include="NUnit.Analyzers" Version="3.9.0"/>
-        <PackageReference Include="NUnit3TestAdapter" Version="4.5.0"/>
+        <PackageReference Include="Avalonia.Headless.NUnit" Version="11.1.3" />
+        <PackageReference Include="Avalonia.ReactiveUI" Version="11.1.0" />
```
Comment (nrdag) [inline DashboardTests/DashboardTests.csproj]: Changes also not required here

File: Dashboard/Views/ConnectionWindow.axaml.cs (hunk 14 above)
Comment (nrdag) [reply Dashboard/Views/ConnectionWindow.axaml.cs]: you didnt fix this still.

File: Dashboard/Views/ConnectionWindow.axaml (hunk 13 above)
Comment (nrdag) [reply Dashboard/Views/ConnectionWindow.axaml]: you didnt fix this

File: Dashboard/Views/ConnectionWindow.axaml (hunk 5 above)
Comment (nrdag) [reply Dashboard/Views/ConnectionWindow.axaml]: you didnt fix this

File: Dashboard/App.axaml.cs (hunk 10 above)
Comment (nrdag) [reply Dashboard/App.axaml.cs]: you didnt do this still

File: Dashboard/Views/ConnectionWindow.axaml [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+                    <ComboBoxItem>File (.CSV)</ComboBoxItem>
+                    
+                </ComboBox>
+
+                <!-- Serial Port -->
+
+                 <TextBlock Text="Serial Port" Margin="0, 10, 0, 5"/>
+                <StackPanel Orientation="Horizontal">
+
+                    <ComboBox x:Name="cmbSerialPort" Width="200" Margin="0, 10, 0, 5"></ComboBox>
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml]: Can you populate this box from the start, so we don't need to click the refresh at start

**Review by nrdag — CHANGES_REQUESTED** [review]
Please ensure you are addressing all comments, this code currently doesn't compile at all. You need to merge in the latest changes from the main branch. You need to ensure every time you ask someone to review a PR, it is ready to be reviewed.

**Comment by Sam6wani** [comment]
Issue fixed related to this ticket. Please review again

File: AV-Datalogger.sln [hunk 20]
```diff
@@ -4,8 +4,6 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Dashboard", "Dashboard\Dash
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "DashboardTests", "DashboardTests\DashboardTests.csproj", "{BA33AD13-B1E8-477D-8601-29E660E6D9BE}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "SerialSimulator", "SerialSimulator\SerialSimulator.csproj", "{C10F65DF-B20D-4EFC-837A-8B5D11AA3ADA}"
```
Comment (nrdag) [inline AV-Datalogger.sln:7]: why was this removed?

File: Dashboard/Connectors/DummyConnector.cs [hunk 21]
```diff
@@ -7,12 +7,11 @@ namespace Dashboard.Connectors;
 
 public class DummyConnector : IConnector
 {
+    private bool _shouldStop;
```
Comment (nrdag) [inline Dashboard/Connectors/DummyConnector.cs:10]: why are there 20+ files of formatting changes unrelated to your pr?

File: Dashboard/Connectors/Serial/SerialConnector.cs [hunk 22]
```diff
@@ -8,13 +8,12 @@ namespace Dashboard.Connectors.Serial;
 
 public partial class SerialConnector(ISerialPort comPort) : IConnector
 {
+    // TODO: Remove below when we have the actual values
```
Comment (nrdag) [inline Dashboard/Connectors/Serial/SerialConnector.cs:11]: this is kept next to the random data generation stuff for a reason, please revert

File: Dashboard/Utils/DependencyInjection.cs [hunk 23]
```diff
@@ -20,7 +20,11 @@ public static IServiceProvider ConfigureServices()
         services.AddSingleton<IDataStore, DataStore>();
         services.AddSingleton<IConnector, SerialConnector>();
 
-        services.AddTransient<MainViewModel>();
+        services.AddTransient<DataViewModel>();
```
Comment (nrdag) [inline Dashboard/Utils/DependencyInjection.cs:23]: probably need to add ur view model here so it can be dependency injected

File: Dashboard/ViewModels/ConnectionWindowViewModel.cs [hunk 24]
```diff
@@ -1,28 +1,130 @@
...
-            set => this.RaiseAndSetIfChanged(ref _selectedConnectionType, value);
+            this.RaiseAndSetIfChanged(ref _selectedConnectionType, value);
+            UpdateVisibility();
         }
+    }
+
+    // Property to hold the available connection types
+    public ObservableCollection<ConnectionType> ConnectionTypes { get; }
+
+    public bool IsSerialPortVisible
```
Comment (nrdag) [inline Dashboard/ViewModels/ConnectionWindowViewModel.cs:65]: can all of these getters not just do something like:
```
    // Enum for Connection Types
    private ConnectionType _selectedConnectionType = ConnectionType.SerialPort;
.....

    public bool IsSerialPortVisible => _selectedConnectionType == ConnectionType.SerialPort;
```

File: Dashboard/Views/ConnectionWindow.axaml.cs [hunk 25]
```diff
@@ -1,40 +1,26 @@
...
 {
     public ConnectionWindow()
     {
         InitializeComponent();
-       
     }
 
 
-    private void InitializeComponent()
+    /*private void InitializeComponent()
```
Comment (nrdag) [inline Dashboard/Views/ConnectionWindow.axaml.cs:13]: Why are there 2 connection pages?

File: Dashboard/ViewModels/ConnectionWindowViewModel.cs [hunk 26]
```diff
@@ -1,28 +1,130 @@
...
+
+public enum ConnectionType
 {
-    internal class ConnectionWindowViewModel : ReactiveObject
+    SerialPort,
+    TCP,
+    File
+}
+
+public class ConnectionWindowViewModel : ReactiveObject
```
Comment (nrdag) [inline Dashboard/ViewModels/ConnectionWindowViewModel.cs:17]: This needs to inherit from view model base, it cannot be a reactive object.

File: Dashboard/ViewModels/MainWindowViewModel.cs [hunk 27]
```diff
@@ -8,108 +8,107 @@
...
+    /// <summary>
+    ///     Simple list template to tie the shorten names to ViewModels and Views
+    ///     ViewModel can be Null as not all Views will require one e.g. Help/About
+    ///     Rendered in order left to right.
+    /// </summary>
+    private readonly List<ListItemTemplate> _templates =
+    [
+        // Currently a bunch of dummy views
+        // When adding to the list use typeof(New Item View/ViewModel).
+        new(typeof(ConnectionView), null, "Connection"),
```
Comment (nrdag) [inline Dashboard/ViewModels/MainWindowViewModel.cs:29]: You need to add your page here, it currently does not display at all when you launch the app

File: Dashboard/ViewModels/MainWindowViewModel.cs [hunk 28]
```diff
@@ -8,108 +8,107 @@
...
+    ///     Checks if the view has already been activated,
+    ///     If not created then create and change.
+    /// </summary>
+    partial void OnSelectedListItemChanged(ListItemTemplate value)
+    {
+        // Check if the view already exist in the dict, if it does not then create it and the appropriate ViewModel as needed.
+        Console.WriteLine("Navigation Bar Item Selection Changed to {0}.", value.Label);
+        if (!_views.ContainsKey(value.Label))
         {
-            // Check if the view already exist in the dict, if it does not then create it and the appropriate ViewModel as needed.
```
Comment (nrdag) [inline Dashboard/ViewModels/MainWindowViewModel.cs:52]: Why was this removed

File: Dashboard/ViewModels/ConnectionWindowViewModel.cs [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+
+    private bool _isTCPVisible;
+
+    // Enum for Connection Types
+    private ConnectionType _selectedConnectionType;
+
+    // Ports
+    private string _selectedPort;
+
+    public ConnectionWindowViewModel()
```
Comment (nrdag) [inline Dashboard/ViewModels/ConnectionWindowViewModel.cs:32]: please fix all warnings
